packages/sysvars/src/slot-history.ts

```diff
@@ -0,0 +1,177 @@
+import {
+    SOLANA_ERROR__CODECS__ENUM_DISCRIMINATOR_OUT_OF_RANGE,
+    SOLANA_ERROR__CODECS__INVALID_BYTE_LENGTH,
+    SOLANA_ERROR__CODECS__INVALID_NUMBER_OF_ITEMS,
+    SolanaError,
+} from '@solana/errors';
+import type { GetAccountInfoApi } from '@solana/rpc-api';
+import type { Rpc } from '@solana/rpc-spec';
+import type { Slot } from '@solana/rpc-types';
+
+import { fetchEncodedSysvarAccount, SYSVAR_SLOT_HISTORY_ADDRESS } from './sysvar';
+
+/**
+ * The `SlotHistory` sysvar.
+ *
+ * A bitvector of slots present over the last epoch.
+ */
+export type SysvarSlotHistory = {
+    bits: bigint[];
+    nextSlot: Slot;
+};
+
+const BITVEC_DISCRIMINATOR = 1;
+// Max number of bits in the bitvector.
+// The Solana SDK defines a constant `MAX_ENTRIES` representing the maximum
+// number of bits that can be represented by the bitvector in the `SlotHistory`
+// sysvar. This value is 1024 * 1024 = 1_048_576.
+// See https://github.com/solana-labs/solana/blob/e0203f22dc83cb792fa97f91dbe6e924cbd08af1/sdk/program/src/slot_history.rs#L43
+const BITVEC_NUM_BITS = 1024 * 1024;
+// The length of the bitvector in blocks.
+// At 64 bits per block, this is 1024 * 1024 / 64 = 16_384.
+const BITVEC_LENGTH = BITVEC_NUM_BITS / 64;
+
+const SLOT_HISTORY_ACCOUNT_DATA_STATIC_SIZE =
+    1 + // Discriminator
+    8 + // bitvector length (u64)
+    BITVEC_LENGTH * 8 +
+    8 + // Number of bits (u64)
+    8; // Next slot (u64)
+
+let memoizedU64Encoder: FixedSizeEncoder<bigint, 8> | undefined;
+let memoizedU64Decoder: FixedSizeDecoder<bigint, 8> | undefined;
+let memoizedU64ArrayEncoder: FixedSizeEncoder<bigint[]> | undefined;
+let memoizedU64ArrayDecoder: FixedSizeDecoder<bigint[]> | undefined;
+
+function getMemoizedU64Encoder(): FixedSizeEncoder<bigint, 8> {
+    if (!memoizedU64Encoder) memoizedU64Encoder = getU64Encoder();
+    return memoizedU64Encoder;
+}
+function getMemoizedU64Decoder(): FixedSizeDecoder<bigint, 8> {
+    if (!memoizedU64Decoder) memoizedU64Decoder = getU64Decoder();
+    return memoizedU64Decoder;
+}
+function getMemoizedU64ArrayEncoder(): FixedSizeEncoder<bigint[]> {
+    if (!memoizedU64ArrayEncoder) memoizedU64ArrayEncoder = getArrayCodec(getU64Codec(), { size: BITVEC_LENGTH });
+    return memoizedU64ArrayEncoder;
+}
+function getMemoizedU64ArrayDecoder(): FixedSizeDecoder<bigint[]> {
+    if (!memoizedU64ArrayDecoder) memoizedU64ArrayDecoder = getArrayCodec(getU64Codec(), { size: BITVEC_LENGTH });
+    return memoizedU64ArrayDecoder;
+}
+
+function getSysvarSlotHistoryEncoder(): Encoder<SysvarSlotHistory> {
+    return createEncoder({
+        fixedSize: SLOT_HISTORY_ACCOUNT_DATA_STATIC_SIZE,
+        write: (value: SysvarSlotHistory, bytes, offset) => {
+            // First byte is the bitvector discriminator.
+            bytes.set([BITVEC_DISCRIMINATOR], offset);
+            offset += 1;
+            // Next 8 bytes are the bitvector length.
+            getMemoizedU64Encoder().write(BigInt(BITVEC_LENGTH), bytes, offset);
```

Comment:
This one is a weird structure though, since there's a few static elements and then one big array of dynamic data. It seems like this is pretty custom, at least for `SlotHistory`.